### PR TITLE
Circuit for NUMBER opcode

### PR DESCRIPTION
**Pull request by bchyl**
1. as with timestamp treatig the number u256 type as u64 tmp.
2. next step，for BlockContext word op such as difficulty should use `RandomLinearCombination<F, 32>` instead of scalar for decompress in bus mapping into v0..v31
4. and maybe also reuse code for BlockContext ops.

### CODE REVIEW

File: zkevm-circuits/src/evm_circuit/witness.rs
Comment (ed255) [inline zkevm-circuits/src/evm_circuit/witness.rs:68]: In the spec, the Block table, in the Number tag contains the RLC enoded value of the block number: https://github.com/appliedzkp/zkevm-specs/blob/657d0b034932ea8fa6515c3ec1f9d1e0055f43b0/src/zkevm_specs/evm/typing.py#L63

I think it's better to assign the value without RLC encoding, like you do in this PR (using the assumption that the number will always fit in the field).  If I remember correctly, Han mentioned that the yellow paper doesn't specify a bound on the block number, but I think 64 bits is a sensible bound, and there have been attempts to make these bounds official: https://eips.ethereum.org/EIPS/eip-1985 (we discussed that we may adhere to this EIP in the zkEVM implementation).

To make the implementation match the spec, I think the spec should be updated so that the block number in the Block table is not RLC encoded.

**Review by ed255 — COMMENTED** [review]
Everything looks good!  Great work!

Nevertheless this implementation is different than the current spec as I have written in a comment.  I think the spec should be updated.

File: zkevm-circuits/src/evm_circuit/witness.rs
Comment (bchyl) [reply zkevm-circuits/src/evm_circuit/witness.rs:68]: Thank you very much for explaining the detailed RLC implementation design background.

Comment (bchyl) [reply zkevm-circuits/src/evm_circuit/witness.rs:68]: Thank you very much for explaining the detailed RLC implementation design background.

**Comment by bchyl** [comment]
> Everything looks good! Great work!
> 
> Nevertheless this implementation is different than the current spec as I have written in a comment. I think the spec should be updated.

updated at [specs pr-140](https://github.com/appliedzkp/zkevm-specs/pull/140). pls help to review, 3gs.

**Review by CPerezz — APPROVED** [review]
LGTM once https://github.com/appliedzkp/zkevm-specs/pull/140 is merged!

Also, probably worth to check the nit on the tests I highlighted :)

File: zkevm-circuits/src/evm_circuit/execution/number.rs
Comment (bchyl) [reply zkevm-circuits/src/evm_circuit/execution/number.rs:99]: nice catch

**Review by ed255 — APPROVED** [review]
LGTM!  Thanks for working on this! :)
Once the github actions pass we can merge this.

**Pull request merged**
